### PR TITLE
Added `adaptive_k` argument to IR Precision metric

### LINKED ISSUE #886: Allow automatically adjusting `k` for `RetrievalPrecision` when `k > num(retrieved documents)`
## 🚀 Feature

Allow `RetrievalPrecision` to use `k=min(k, preds.shape[-1])` for individual queries.

### Motivation

When one calculates `precision at k` for a query over the list of documents `D`, where `len(D) < k`, there's an ambiguity whether `k` should be fixed or adjusted w.r.t. `len(D)` when `D < k`.

### Pitch

Add an input argument for an option of how to handle these edge cases. Set a default to match the current implementation.

### Alternatives

Leave it as it is.

### Additional context

The implementation described above is used by the [`catboost` package](https://github.com/catboost/catboost).

### PR DESCRIPTION
**Pull request by lucadiliello**
## What does this PR do?

Fixes #886

## Before submitting

- [X] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [X] Did you make sure to **update the docs**?
- [X] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/retrieval/test_precision.py [hunk 1]
```diff
@@ -43,7 +45,7 @@ def _precision_at_k(target: np.ndarray, preds: np.ndarray, k: int = None):
     assert target.shape == preds.shape
     assert len(target.shape) == 1  # works only with single dimension inputs
 
-    if k is None:
+    if k is None or (adaptive_k and k > len(preds)):
```
Comment (Borda) [inline tests/retrieval/test_precision.py:48]:
```suggestion
    if k is None or adaptive_k and k > len(preds):
```

**Review by Borda — COMMENTED** [review]
pls add changelog entry

File: torchmetrics/retrieval/precision.py [hunk 2]
```diff
@@ -81,6 +84,7 @@ def __init__(
         empty_target_action: str = "neg",
         ignore_index: Optional[int] = None,
         k: Optional[int] = None,
+        adaptive_k: Optional[bool] = False,
```
Comment (SkafteNicki) [inline torchmetrics/retrieval/precision.py:87]:
```suggestion
        adaptive_k: bool = False,
```

**Review by SkafteNicki — COMMENTED** [review]
I assume that the parameter should also be added to the modular metrics

**Comment by lucadiliello** [comment]
LGTM

**Review by stancld — APPROVED** [review]
LGTM

**Comment by lucadiliello** [comment]
> > I assume that the parameter should also be added to the modular metrics
> 
> @lucadiliello not sure if I understand, so does the parameter shall also be added to the modular metrics?

I did add it both to the functional and the modular version of `RetrievalPrecision`. I didn’t add it to the other retrieval metrics because it makes no sense.

File: CHANGELOG.md [hunk 3]
```diff
@@ -44,6 +44,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Added negative `ignore_index` for the Accuracy metric ([#362](https://github.com/PyTorchLightning/metrics/pull/362))
 
 
+- Added `adaptive_k` for the RetrievalPrecision metric ([#910](https://github.com/PyTorchLightning/metrics/pull/910))
```
Comment (Borda) [inline CHANGELOG.md:47]:
```suggestion
- Added `adaptive_k` for the `RetrievalPrecision` metric ([#910](https://github.com/PyTorchLightning/metrics/pull/910))
```

**Pull request merged**